### PR TITLE
feat: support for many padding dynamo converters

### LINKED ISSUE #2461: Add support for padding aten operators
```[tasklist]
### Tasks
- [x] torch.ops.aten.pad.default
- [x] torch.ops.aten.constant_pad_nd.default
- [x] torch.ops.aten.reflection_pad1d.default
- [x] torch.ops.aten.reflection_pad2d.default
- [x] torch.ops.aten.reflection_pad3d.default
- [x] torch.ops.aten.replication_pad1d.default
- [x] torch.ops.aten.replication_pad2d.default
- [x] torch.ops.aten.replication_pad3d.default
- [x] torch.ops.aten._pad_circular.default
```

### PR DESCRIPTION
**Pull request by zewenli98**
# Description

Support for many padding dynamo converters, including:

```
torch.ops.aten.pad.default
torch.ops.aten.constant_pad_nd.default
torch.ops.aten.reflection_pad1d.default
torch.ops.aten.reflection_pad2d.default
torch.ops.aten.reflection_pad3d.default
torch.ops.aten.replication_pad1d.default
torch.ops.aten.replication_pad2d.default
torch.ops.aten.replication_pad3d.default
torch.ops.aten._pad_circular.default
```

Note: 
1. Not sure why there're no similar names for `torch.ops.aten._pad_circular.default` like `torch.ops.aten.circular_pad1d.default`, but according to [PyTorch doc](https://pytorch.org/docs/stable/generated/torch.nn.functional.pad.html#torch.nn.functional.pad), it does have the similar functionality as `reflection` and `replication`, targeting for 2D, 3D, 4D, and 5D inputs.
2. `torch.ops.aten.pad.default` is a higher-level function to call the other padding functions.

Fixes #2461 #2211 

## Type of change

- New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code follows the style guidelines of this project (You can use the linters)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes
- [x] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: py/torch_tensorrt/dynamo/conversion/impl/slice/ops.py [hunk 1]
```diff
@@ -37,10 +37,6 @@ def slice_op(  # TODO: This should be slice not whatever is in base
             "of the TensorRT region!"
         )
 
-    dim = get_positive_dim(dim, len(input.shape))
-    start = get_positive_dim(start, input.shape[dim])
-    stop = get_positive_dim(stop, input.shape[dim])
-
```
Comment (zewenli98) [inline py/torch_tensorrt/dynamo/conversion/impl/slice/ops.py:43]: I did this change because in some cases I want to use `slice_op` to get a reserved tensor but the `stop` arg cannot be negative. For example, when doing:
```
impl.slice.slice_op(
                ctx,
                target,
                source_ir,
                f"{name}_slice_post{i}",
                input,
                dim=dim,
                start=2,
                stop=-1,
                step=-1,
            )
```
if has these lines, `stop` would be a positive value (i.e. `length - 1`), but since `step` is still negative, it will never stop, however, my expected sequence is `[2, 1, 0]`. 

The reason why I don't use positive `stop` and `step` is I wanna get a reversed tensor.

The unit tests proved these lines are not necessary.

File: py/torch_tensorrt/dynamo/conversion/impl/pad.py [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+
+    # Implement constant padding via concat
+    curr_dim = len(input.shape) - 1
+
+    for i in range(0, len(pad), 2):
+        input_shape = list(input.shape)
+
+        pre_pad = pad[i]
+        post_pad = pad[i + 1]
+        pre_pad_shape = copy.deepcopy(input_shape)
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/impl/pad.py:38]: Is the `list` deepcopy required here? Could we instead use `tuple`s to avoid the need for `copy.deepcopy`?

File: py/torch_tensorrt/dynamo/conversion/impl/pad.py [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+        pre_pad_shape[curr_dim] = pre_pad
+        pre_pad_tensor = torch.full(pre_pad_shape, float(value))
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/impl/pad.py:40]: Due to the `FakeTensor` issue in `torch.compile`, could this instead be a Numpy array?

File: py/torch_tensorrt/dynamo/conversion/impl/pad.py [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+    name: str,
+    input: TRTTensor,
+    padding: Sequence[int],
+) -> TRTTensor:
+    if has_dynamic_shape(input.shape):
+        assert input.shape[1] != -1, "Channel dim can't be dynamic for padding."
+
+    padding_dims = len(padding) // 2
+
+    if padding_dims == 1 or padding_dims == 2 or padding_dims == 3:
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/impl/pad.py:74]: Consider condensing to `if padding_dims in (1, 2, 3)`

File: py/torch_tensorrt/dynamo/conversion/impl/slice/ops.py (hunk 1 above)
Comment (gs-olive) [reply py/torch_tensorrt/dynamo/conversion/impl/slice/ops.py:43]: The issue with removing this is that the line `output_shape[dim] = math.ceil((stop - start) / step)` below will be incorrect when certain values are negative. This change would have to be accompanied by a fix for that equation, since these test cases should otherwise fail:
https://github.com/pytorch/TensorRT/blob/867dc7bf23b34f8d3809a1d64841b0013a9a0054/tests/py/dynamo/conversion/test_slice_aten.py#L16-L17

Comment (zewenli98) [reply py/torch_tensorrt/dynamo/conversion/impl/slice/ops.py]: Thanks for the comments! I undid the change due to better implementation.

**Comment by zewenli98** [comment]
Hi @gs-olive, thanks for your review! I just found out a better way to implement paddings, with the help of `ISliceLayer` with different `SliceMode`. So I undid the change to `slice_op`.

File: py/torch_tensorrt/dynamo/conversion/impl/pad.py [hunk 5]
```diff
@@ -1,14 +1,22 @@
...
 
-import torch
-import torch_tensorrt.dynamo.conversion.impl as impl
+import tensorrt as trt
 from torch.fx.node import Target
 from torch_tensorrt.dynamo._SourceIR import SourceIR
 from torch_tensorrt.dynamo.conversion._ConversionContext import ConversionContext
-from torch_tensorrt.fx.converters.converter_utils import has_dynamic_shape
+from torch_tensorrt.fx.converters.converter_utils import (
+    get_trt_tensor,
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/impl/pad.py:8]: This should use the `dynamo.converters.converter_utils` implementation of `get_trt_tensor`

File: py/torch_tensorrt/dynamo/conversion/impl/pad.py [hunk 6]
```diff
@@ -19,43 +27,36 @@ def constant_padNd(
...
-        output = impl.cat.cat(
-            ctx,
-            target,
-            source_ir,
-            f"{name}_concat{curr_dim}",
-            input=(pre_pad_tensor, input, post_pad_tensor),
-            dim=curr_dim,
+    rank = len(input.shape)
+
+    if len(pad) / 2 > rank:
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/impl/pad.py:35]: Should this be a `//` instead of `/`?

Comment (zewenli98) [reply py/torch_tensorrt/dynamo/conversion/impl/pad.py:35]: Both are fine because `len(pad)` should be even. BTW, if PyTorch checks some restrictions like `len(pad) should be even`, do we need to check again within our converters?

Comment (gs-olive) [reply py/torch_tensorrt/dynamo/conversion/impl/pad.py:35]: It shouldn't be necessary if PyTorch checks it, but can't hurt to add the check

Comment (gs-olive) [reply py/torch_tensorrt/dynamo/conversion/impl/pad.py:35]: Additionally, `//` is preferred if the results shouldn't matter, since `//` will output an integer

File: py/torch_tensorrt/dynamo/conversion/impl/pad.py [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+
+    rank = len(input.shape)
+
+    if len(pad) // 2 > rank:
+        raise RuntimeError(
+            f"Trying to pad last {len(pad) // 2} dimension but the input only has {rank} dimension."
+        )
+
+    start_list = [0] * len(input.shape)
+    new_shape = input.shape
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/impl/pad.py:157]: Since `new_shape` gets assigned to below, I think this might need to be `list(input.shape)`, to ensure it is casted to a mutable list.

Comment (gs-olive) [reply py/torch_tensorrt/dynamo/conversion/impl/pad.py:157]: This applies to the other cases above as well. It may not be required with TRT's current implementation of `input.shape`, but it would be preferable in general to have this cast to be safe.

File: py/torch_tensorrt/dynamo/conversion/impl/pad.py [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+
+    for i in range(0, len(padding) // 2):
+        start_list[-i - 1] = -padding[i * 2]
+        new_shape[-i - 1] += padding[i * 2] + padding[i * 2 + 1]
+
+    stride_list = [1] * len(new_shape)
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/impl/pad.py:125]: For clarity, could `len(input.shape)` on line 118 and `len(new_shape)` on line 125 both be replaced with `rank`, or would this make the code incorrect?

The above applies to a few other functions as well I believe.

**Review by gs-olive — COMMENTED** [review]
Overall looks great, added a few style comments and suggested modifications.

File: py/torch_tensorrt/dynamo/conversion/impl/pad.py (hunk 7 above)
Comment (zewenli98) [reply py/torch_tensorrt/dynamo/conversion/impl/pad.py:157]: That makes sense, thanks for the suggestion!

**Review by gs-olive — APPROVED** [review]
Looks good to me!
